server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v71/MigrateWebhooksToWebhooksTable.java

```diff
@@ -0,0 +1,316 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.platform.db.migration.version.v71;
+
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+import org.apache.commons.lang.StringUtils;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+import org.sonar.core.util.UuidFactory;
+import org.sonar.db.Database;
+import org.sonar.server.platform.db.migration.step.DataChange;
+import org.sonar.server.platform.db.migration.step.Upsert;
+import org.sonar.server.platform.db.migration.version.v63.DefaultOrganizationUuidProvider;
+
+import static java.util.Collections.emptyList;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toMap;
+
+public class MigrateWebhooksToWebhooksTable extends DataChange {
+
+  private static final Logger LOGGER = Loggers.get(MigrateWebhooksToWebhooksTable.class);
+
+  private DefaultOrganizationUuidProvider defaultOrganizationUuidProvider;
+  private UuidFactory uuidFactory;
+
+  public MigrateWebhooksToWebhooksTable(Database db, DefaultOrganizationUuidProvider defaultOrganizationUuidProvider, UuidFactory uuidFactory) {
+    super(db);
+    this.defaultOrganizationUuidProvider = defaultOrganizationUuidProvider;
+    this.uuidFactory = uuidFactory;
+  }
+
+  @Override
+  public void execute(Context context) throws SQLException {
+
+    Map<String, PropertyRow> rows = context
+      .prepareSelect("select id, prop_key, resource_id, text_value, created_at from properties")
+      .list(row -> new PropertyRow(
+        row.getLong(1),
+        row.getString(2),
+        row.getLong(3),
+        row.getString(4),
+        row.getLong(5)))
+      .stream()
+      .filter(row -> row.key().startsWith("sonar.webhooks"))
+      .collect(toMap(PropertyRow::key, Function.identity()));
+
+    if (!rows.isEmpty()) {
+      final List<Long> toBeDeleted = new ArrayList<>();
+      toBeDeleted.addAll(migrateGlobalWebhooks(context, rows));
+      toBeDeleted.addAll(migrateProjectsWebhooks(context, rows, getProjectsUuids(context)));
+      deleteFromProperties(context, toBeDeleted);
+    }
+
+  }
+
+  private static Map<Long, String> getProjectsUuids(Context context) throws SQLException {
+    return context
+      .prepareSelect("select id, uuid from projects")
```

Comment:
This table is huge and should not be loaded in memory. Loading project one by one on demand is probably fine in the current state of webhooks. I don't expect to have so many webhooks declared in an instance of SQ.

Reply:
done